src/Compile.hs

```diff
@@ -28,85 +28,98 @@ class Compilable a where
+      -- Solved _ _ (App (Solved _ _ (Var "==")) arg') ->
+      --   compile astPath outputPath arg'
+      --     <> " === "
+      --     <> compile astPath outputPath arg
+      Solved _ _ (App (Solved _ _ (Var "&&")) arg' _) ->
         compile astPath outputPath arg'
           <> " && "
           <> compile astPath outputPath arg
-      Solved _ _ (App (Solved _ _ (Var "||")) arg') ->
+      Solved _ _ (App (Solved _ _ (Var "||")) arg' _) ->
         compile astPath outputPath arg'
           <> " || "
           <> compile astPath outputPath arg
-      Solved _ _ (App (Solved _ _ (Var ">")) arg') ->
+      Solved _ _ (App (Solved _ _ (Var ">")) arg' _) ->
         compile astPath outputPath arg'
           <> " > "
           <> compile astPath outputPath arg
-      Solved _ _ (App (Solved _ _ (Var "<")) arg') ->
+      Solved _ _ (App (Solved _ _ (Var "<")) arg' _) ->
         compile astPath outputPath arg'
           <> " < "
           <> compile astPath outputPath arg
-      Solved _ _ (App (Solved _ _ (Var ">=")) arg') ->
+      Solved _ _ (App (Solved _ _ (Var ">=")) arg' _) ->
         compile astPath outputPath arg'
           <> " >= "
           <> compile astPath outputPath arg
-      Solved _ _ (App (Solved _ _ (Var "<=")) arg') ->
+      Solved _ _ (App (Solved _ _ (Var "<=")) arg' _) ->
         compile astPath outputPath arg'
           <> " <= "
           <> compile astPath outputPath arg
 
-      Solved _ _ (App (Solved _ _ (Var "|>")) arg') ->
+      Solved _ _ (App (Solved _ _ (Var "|>")) arg' _) ->
         compile astPath outputPath arg
           <> "("
           <> compile astPath outputPath arg'
           <> ")"
 
-      _ -> compileApp [] abs arg
+      _ -> compileApp [] [] abs arg final
      where
-      compileApp :: [String] -> Exp -> Exp -> String
-      compileApp prevArgs abs arg =
-        let
-          args = [compile astPath outputPath arg] <> prevArgs
-          next = case abs of
-            (Solved _ _ (App abs' arg')) -> compileApp args abs' arg'
-            _ ->
-              compile astPath outputPath abs
-                <> "("
-                <> intercalate ", " args
-                <> ")"
-        in
-          next
+      compileApp :: [Bool] -> [String] -> Exp -> Exp -> Bool -> String
+      compileApp prevFinals prevArgs abs arg final =
+        let args   = [compile astPath outputPath arg] <> prevArgs
+            finals = [final] <> prevFinals
+            next   = case abs of
+              (Solved _ _ (App abs' arg' final')) ->
+                compileApp finals args abs' arg' final'
+              _ ->
+                let finalized = zip args finals
+                in  compile astPath outputPath abs
+                      <> "("
+                      <> buildParams finalized
+                      <> ")"
+        in  next
+
+      buildParams :: [(String, Bool)] -> String
+      buildParams []                    = ""
+      buildParams ((arg, final) : args) = if final
```

Comment:
And that's the thing doing the recursion and deciding what to do depending on if it's a final arg or not.